# Predicting sentiment from product reviews

## Fire up GraphLab Create

In [53]:
import graphlab

# Read some product review data

Loading reviews for a set of baby products. 

In [54]:
products = graphlab.SFrame('amazon_baby.gl/')

# Let's explore this data together

Data includes the product name, the review text and the rating of the review. 

In [55]:
products.head()

Columns:
	name	str
	review	str
	rating	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
[10 rows x 3 columns]

# Build the word count vector for each review

In [56]:
products['word_count'] = graphlab.text_analytics.count_words(products['review'])

In [57]:
products.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+
|           word_count          |
+-------------------------------+
| {'and': 5, 'stink': 1, 'be... |
| {'and': 3, 'love': 1, 'it'... |
| {'and': 2, 'quilt': 1, 'it... |
| {'ingenious': 1, 'and': 3,... |
| {'and': 2, 'parents!!': 1,... |
| {'and': 2, 'cute': 1, 'hel... |
| {'shop': 1, 'be': 1, 'is':... |
| {'feeding,': 1, 'and': 2, ... |
| {'and': 1, 'help': 1, 'giv... |
| {'journal.': 1, 'all': 1, ... |
+-------------------------------+
[10 rows x 4 columns]

In [58]:
graphlab.canvas.set_target('ipynb')

In [59]:
products['name'].show()

# Examining the reviews for most-sold product:  'Vulli Sophie the Giraffe Teether'

In [60]:
giraffe_reviews = products[products['name'] == 'Vulli Sophie the Giraffe Teether']

In [61]:
len(giraffe_reviews)

785

In [62]:
giraffe_reviews['rating'].show(view='Categorical')

# Build a sentiment classifier

In [63]:
products['rating'].show(view='Categorical')

## Define what's a positive and a negative sentiment

We will ignore all reviews with rating = 3, since they tend to have a neutral sentiment.  Reviews with a rating of 4 or higher will be considered positive, while the ones with rating of 2 or lower will have a negative sentiment.   

In [64]:
#ignore all 3* reviews
products = products[products['rating'] != 3]

In [65]:
#positive sentiment = 4* or 5* reviews
products['sentiment'] = products['rating'] >=4

In [66]:
products.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+
|           word_count          | sentiment |
+-------------------------------+-----------+
| {'and': 3, 'love': 1, 'it'... |     1     |
| {'and': 2, 'quilt': 1, 'it... |     1     |
| {'ingenious': 1, 'and': 3,... |     1     |
| {'and': 2, 'parents!!': 1,... |     1     |
| {'and': 2, 'cute': 1, 'hel... |     1     |
| {'shop': 1, 'be': 1, 'is':... |     1     |
| {'feeding,': 1, 'and': 2, ... |     1     |
| {'and': 1, 'help': 1, 'giv... |     1     |
| {'journal.': 1, 'all': 1, ... |     1     |
| {'all': 1, 'forget': 1, 'j... |     1     |
+-------------------------------+-----------+
[10 rows x 5 columns]

## Let's train the sentiment classifier

In [67]:
train_data,test_data = products.random_split(.8, seed=0)

In [68]:
sentiment_model = graphlab.logistic_classifier.create(train_data,
                                                     target='sentiment',
                                                     features=['word_count'],
                                                     validation_set=test_data)

PROGRESS: Logistic regression:
PROGRESS: --------------------------------------------------------
PROGRESS: Number of examples          : 133448
PROGRESS: Number of classes           : 2
PROGRESS: Number of feature columns   : 1
PROGRESS: Number of unpacked features : 219217
PROGRESS: Number of coefficients    : 219218
PROGRESS: Starting L-BFGS
PROGRESS: --------------------------------------------------------
PROGRESS: +-----------+----------+-----------+--------------+-------------------+---------------------+
PROGRESS: | Iteration | Passes   | Step size | Elapsed Time | Training-accuracy | Validation-accuracy |
PROGRESS: +-----------+----------+-----------+--------------+-------------------+---------------------+
PROGRESS: | 1         | 5        | 0.000002  | 3.663895     | 0.841481          | 0.839989            |
PROGRESS: | 2         | 9        | 3.000000  | 5.892106     | 0.947425          | 0.894877            |
PROGRESS: | 3         | 10       | 3.000000  | 6.736414     | 0.92

# Evaluate the sentiment model

In [101]:
sentiment_model.evaluate(test_data, metric='accuracy')

{'accuracy': 0.916256305548883}

In [69]:
sentiment_model.evaluate(test_data, metric='roc_curve')

{'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 1001
 
 Data:
 +------------------+----------------+------------------+-------+------+
 |    threshold     |      fpr       |       tpr        |   p   |  n   |
 +------------------+----------------+------------------+-------+------+
 |       0.0        | 0.222410227486 | 0.00448829836498 | 28073 | 5319 |
 | 0.0010000000475  | 0.777589772514 |  0.995511701635  | 28073 | 5319 |
 | 0.00200000009499 | 0.738296672307 |  0.994300573505  | 28073 | 5319 |
 | 0.00300000002608 | 0.715736040609 |  0.993623766608  | 28073 | 5319 |
 | 0.00400000018999 | 0.700319608949 |  0.99312506679   | 28073 | 5319 |
 | 0.00499999988824 |  0.6890392931  |  0.992733231219  | 28073 | 5319 |
 | 0.00600000005215 |  0.6790750141  |   0.9922345314   | 28073 | 5319 |
 | 0.00700000021607 | 0.668734724572 |  0.991842695829  | 28073 | 5319 |
 | 0.00800000037998 | 0.658582440308 |  0.991522103088  | 28073 | 5319 |
 | 0.00899999961257 

In [70]:
sentiment_model.show(view='Evaluation')

# Applying the learned model to understand sentiment for Giraffe

In [71]:
giraffe_reviews['predicted_sentiment'] = sentiment_model.predict(giraffe_reviews, output_type='probability')

In [72]:
giraffe_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | He likes chewing on all th... |  5.0   |
| Vulli Sophie the Giraffe T... | My son loves this toy and ... |  5.0   |
| Vulli Sophie the Giraffe T... | There really should be a l... |  1.0   |
| Vulli Sophie the Giraffe T... | All the moms in my moms' g... |  5.0   |
| Vulli Sophie the Giraffe T... | I was a little skeptical o... |  5.0   |
| Vulli Sophie the Giraffe T... | I have been reading about ... |  5.0   |
| Vulli Sophie the Giraffe T... | My neice loves her sophie ... |  5.0   |
| Vulli Sophie the Giraffe T... | What a friendly face!  And... |  5.0   |
| Vulli Sophie the Giraffe T... | We got this just for my so... |  5.0   |
| Vulli Sophie the Giraffe T... | My baby seems to like this... |  3.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+---------------------+
|           word_count          | predicted_sentiment |
+-------------------------------+---------------------+
| {'and': 1, 'all': 1, 'beca... |    0.999513023521   |
| {'and': 1, 'right': 1, 'he... |    0.999320678306   |
| {'and': 2, 'all': 1, 'late... |    0.013558811687   |
| {'and': 2, 'one!': 1, 'all... |    0.995769474148   |
| {'and': 3, 'all': 1, 'old'... |    0.662374415673   |
| {'and': 6, 'seven': 1, 'al... |    0.999997148186   |
| {'and': 4, 'drooling,': 1,... |    0.989190989536   |
| {'and': 3, 'chew': 1, "don... |    0.999563518413   |
| {'chew': 2, 'because': 1, ... |    0.970160542725   |
| {'and': 2, 'already': 1, '... |    0.195367644588   |
+-------------------------------+---------------------+
[10 rows x 5 columns]

## Sort the reviews based on the predicted sentiment and explore

In [73]:
giraffe_reviews = giraffe_reviews.sort('predicted_sentiment', ascending=False)

In [74]:
giraffe_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | Sophie, oh Sophie, your ti... |  5.0   |
| Vulli Sophie the Giraffe T... | I'm not sure why Sophie is... |  4.0   |
| Vulli Sophie the Giraffe T... | I'll be honest...I bought ... |  4.0   |
| Vulli Sophie the Giraffe T... | We got this little giraffe... |  5.0   |
| Vulli Sophie the Giraffe T... | As a mother of 16month old... |  5.0   |
| Vulli Sophie the Giraffe T... | Sophie the Giraffe is the ... |  5.0   |
| Vulli Sophie the Giraffe T... | Sophie la giraffe is absol... |  5.0   |
| Vulli Sophie the Giraffe T... | My 5-mos old son took to t... |  5.0   |
| Vulli Sophie the Giraffe T... | My nephews and my four kid... |  5.0   |
| Vulli Sophie the Giraffe T... | Never thought I'd see my s... |  5.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+---------------------+
|           word_count          | predicted_sentiment |
+-------------------------------+---------------------+
| {'giggles': 1, 'all': 1, "... |         1.0         |
| {'peace': 1, 'month': 1, '... |    0.999999999703   |
| {'all': 2, 'pops': 1, 'exi... |    0.999999999392   |
| {'all': 2, "don't": 1, '(l... |    0.99999999919    |
| {'cute': 1, 'all': 1, 'rev... |    0.999999998657   |
| {'just': 2, 'both': 1, 'mo... |    0.999999997108   |
| {'and': 5, 'the': 1, 'all'... |    0.999999995589   |
| {'just': 1, 'shape': 2, 'm... |    0.999999995573   |
| {'and': 4, 'chew': 1, 'all... |    0.999999989527   |
| {'giggles': 1, 'all': 1, '... |    0.999999985069   |
+-------------------------------+---------------------+
[10 rows x 5 columns]

##Most positive reviews for the giraffe

In [75]:
giraffe_reviews[0]['review']

"Sophie, oh Sophie, your time has come. My granddaughter, Violet is 5 months old and starting to teeth. What joy little Sophie brings to Violet. Sophie is made of a very pliable rubber that is sturdy but not tough. It is quite easy for Violet to twist Sophie into unheard of positions to get Sophie into her mouth. The little nose and hooves fit perfectly into small mouths, and the drooling has purpose. The paint on Sophie is food quality.Sophie was born in 1961 in France. The maker had wondered why there was nothing available for babies and made Sophie from the finest rubber, phthalate-free on St Sophie's Day, thus the name was born. Since that time millions of Sophie's populate the world. She is soft and for babies little hands easy to grasp. Violet especially loves the bumpy head and horns of Sophie. Sophie has a long neck that easy to grasp and twist. She has lovely, sizable spots that attract Violet's attention. Sophie has happy little squeaks that bring squeals of delight from Viol

In [76]:
giraffe_reviews[1]['review']

"I'm not sure why Sophie is such a hit with the little ones, but my 7 month old baby girl is one of her adoring fans.  The rubber is softer and more pleasant to handle, and my daughter has enjoyed chewing on her legs and the nubs on her head even before she started teething.  She also loves the squeak that Sophie makes when you squeeze her.  Not sure what it is but if Sophie is amongst a pile of her other toys, my daughter will more often than not reach for Sophie.  And I have the peace of mind of knowing that only edible and safe paints and materials have been used to make Sophie, as opposed to Bright Starts and other baby toys made in China.  Now that the research is out on phthalates and other toxic substances in baby toys, I think it's more important than ever to find good quality toys that are also safe for our babies to handle and put in their mouths.  Sophie is a must-have for every new mom in my opinion.  Even if your kid is one of the few that can take or leave her, it's worth

## Show most negative reviews for giraffe

In [77]:
giraffe_reviews[-1]['review']

"My son (now 2.5) LOVED his Sophie, and I bought one for every baby shower I've gone to. Now, my daughter (6 months) just today nearly choked on it and I will never give it to her again. Had I not been within hearing range it could have been fatal. The strange sound she was making caught my attention and when I went to her and found the front curved leg shoved well down her throat and her face a purply/blue I panicked. I pulled it out and she vomited all over the carpet before screaming her head off. I can't believe how my opinion of this toy has changed from a must-have to a must-not-use. Please don't disregard any of the choking hazard comments, they are not over exaggerated!"

In [78]:
giraffe_reviews[-2]['review']

"This children's toy is nostalgic and very cute. However, there is a distinct rubber smell and a very odd taste, yes I tried it, that my baby did not enjoy. Also, if it is soiled it is extremely difficult to clean as the rubber is a kind of porus material and does not clean well. The final thing is the squeaking device inside which stopped working after the first couple of days. I returned this item feeling I had overpaid for a toy that was defective and did not meet my expectations. Please do not be swayed by the cute packaging and hype surounding it as I was. One more thing, I was given a full refund from Amazon without any problem."

# Quiz

Instead of using all the words in the reviews, only a subset of words will be used.

In [79]:
selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']

##  Use .apply() to build a new feature with the counts for each of the selected_words

### create a column products[‘awesome’] where each row contains the number of times the word ‘awesome’ showed up in the review for the corresponding product, and 0 if the review didn’t show up

One way to do this is to look at the each row ‘word_count’ column and follow this logic:

 *   If ‘awesome’ shows up in the word counts for a particular product (row of the products SFrame), then we know how often ‘awesome’ appeared in the review, 

 *  if ‘awesome’ doesn’t appear in the word counts, then it didn’t appear in the review, and we should set the count for ‘awesome’ to 0 in this review. 

#### write a function called `awesome_count` which takes in the word counts and returns the number of times ‘awesome’ appears in the reviews.

 *  Each entry of the ‘word_count’ column is of Python type dictionary.



In [80]:
def awesome_count(word_counts):
    return word_counts.get('awesome', 0)

#### use .apply() to iterate awesome_count for each row of products[‘word_count’] and create a new column called ‘awesome’ with the resulting counts

In [81]:
products['awesome'] = products['word_count'].apply(awesome_count)

#### Repeat this process for the other 11 words in selected_words.

In [82]:
for word in selected_words:
    products[word] = products['word_count'].apply(lambda wc: wc.get(word, 0))

### Using the .sum() method on each of the new columns you created, answer the following questions

#### Out of the selected_words, which one is most used in the dataset?

In [83]:
max((products[word].sum(), word) for word in selected_words)

(42420, 'great')

####  Which one is least used?

In [84]:
min((products[word].sum(), word) for word in selected_words)

(131, 'wow')

##  Create a new sentiment analysis model using only the selected_words as features

### Split the data into testing and training sets

In [85]:
train_data,test_data = products.random_split(.8, seed=0)

### Train a logistic regression classifier (use graphlab.logistic_classifier.create) using just the selected_words.

In [86]:
 selected_words_model = graphlab.logistic_classifier.create(train_data,
                                                     target='sentiment',
                                                     features=selected_words,
                                                     validation_set=test_data)

PROGRESS: Logistic regression:
PROGRESS: --------------------------------------------------------
PROGRESS: Number of examples          : 133448
PROGRESS: Number of classes           : 2
PROGRESS: Number of feature columns   : 11
PROGRESS: Number of unpacked features : 11
PROGRESS: Number of coefficients    : 12
PROGRESS: Starting Newton Method
PROGRESS: --------------------------------------------------------
PROGRESS: +-----------+----------+--------------+-------------------+---------------------+
PROGRESS: | Iteration | Passes   | Elapsed Time | Training-accuracy | Validation-accuracy |
PROGRESS: +-----------+----------+--------------+-------------------+---------------------+
PROGRESS: | 1         | 2        | 0.322597     | 0.844299          | 0.842842            |
PROGRESS: | 2         | 3        | 0.537594     | 0.844186          | 0.842842            |
PROGRESS: | 3         | 4        | 0.748692     | 0.844276          | 0.843142            |
PROGRESS: | 4         | 5        |

##  sort the learned coefficients according to the ‘value’ column using .sort()

In [91]:
swm =  selected_words_model['coefficients'].sort('value')
swm.show()

#### What word had the most positive weight?

In [96]:
print('Most Positive Word: "{0}" ({1})'.format(swm[-1]['name'],
                                            swm[-1]['value']))

Most Positive Word: "love" (1.39989834302)


#### What word had the most negative weight?

In [97]:
print('Most Negative Word: "{0}" ({1})'.format(swm[0]['name'],
                                              swm[0]['value']))

Most Negative Word: "terrible" (-2.09049998487)


This makes sense since the classifier is trying to predict a rating of 4 or more, so it should give more weight to a positive word ('love') and less weight to a negative word ('terrible').

## Compare the accuracy of the sentiment models

### Selected Words Model Accuracy

In [99]:
selected_words_model.evaluate(test_data, metric='accuracy')

{'accuracy': 0.8431419649291376}

### Sentiment Model Accuracy

In [102]:
sentiment_model.evaluate(test_data, metric='accuracy')

{'accuracy': 0.916256305548883}

### Majority Class Accuracy

In [103]:
total = len(products)
positive = sum(products['sentiment'])
negative = total - positive
assert positive > negative
print("Positive Sentiment is the majority sentiment")

Positive Sentiment is the majority sentiment


In [104]:
total = len(test_data)
positive = sum(test_data['sentiment'])
print("Accuracy if you always predict a positive sentiment: {0}".format(positive/float(total)))

Accuracy if you always predict a positive sentiment: 0.840019216911


Using all the words provides a greater accuracy than the baseline model of always predicting that a sentiment is positive, while the selected words model was only negligibly better.

## Interpreting the Performance differences

### Isolate the 'Baby Trend Diaper Champ' product

In [105]:
diaper_champ_reviews = products[products['name'] == 'Baby Trend Diaper Champ']
diaper_champ_reviews.show()

In [106]:
diaper_champ_reviews['predicted_sentiment'] = sentiment_model.predict(diaper_champ_reviews, output_type='probability')

In [107]:
diaper_champ_reviews = diaper_champ_reviews.sort('predicted_sentiment', ascending=False)

### What is the predicted sentiment for the most positive review (using the sentiment_model)?

In [117]:
# to get the predicted_sentiment you can just select diaper_champ_reviews[0]
# but the `predict` method needs to use [0:1] as the slice
predicted_most_positive = diaper_champ_reviews[0:1]
predicted_most_positive['predicted_sentiment']

dtype: float
Rows: 1
[0.9999999372669541]

### What does the selected_words_model predict for that same review?

In [118]:
selected_words_model.predict(predicted_most_positive, output_type='probability')

dtype: float
Rows: 1
[0.7969408512906712]

### Why is the sentiment for the sentiment_model more positive than the selected_words_model?

In [121]:
word_counts = predicted_most_positive['word_count'][0]
sum(word_counts[word] for word in word_counts if word in selected_words)

0

None of the selected words were in the review.

In [122]:
sum(v for k,v in word_counts.iteritems())

173

In [123]:
len(word_counts)

114

So the `selected_words_model` didn't use any of the 114 words in the review and thus missed out on its 173 points.